### PR TITLE
add alert for high cpu usgage

### PR DESCRIPTION
**Pull request by austincunningham**
# Description

Jira: https://issues.redhat.com/browse/INTLY-7234

Adds alerts for external redis cpu high usage

# Verification
## Verify Deploy
- Set the useClusterStorage var to `false` in the `deploy/crds/examples/rhmi.cr.yaml`
```yaml
apiVersion: integreatly.org/v1alpha1
kind: RHMI
metadata:
  name: example-rhmi
spec:
  type: managed
  namespacePrefix: redhat-rhmi-
  selfSignedCerts: true
  useClusterStorage: 'false'
  smtpSecret: redhat-rhmi-smtp
  deadMansSnitchSecret: redhat-rhmi-deadmanssnitch
  pagerDutySecret: redhat-rhmi-pagerduty
```
- Run `make cluster/prepare/local`
- Run `oc create -f deploy/crds/examples/rhmi.cr.yaml`
- Run `make code/run`
- Once 3scale has finished deploying check for the following alerts
 ``` 
RedisCpuUsageHigh
```
image
image

## Verify Alert will trigger
- Load the memory to 75% following this readme https://github.com/integr8ly/cloud-resource-operator/blob/master/hack/redis/README.md in the cloud resources operator
- Load the Cpu to 100% following the same readme https://github.com/integr8ly/cloud-resource-operator/blob/master/hack/redis/README.md in the cloud resources operator
- Wait 15min for alert to trigger.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist
- [ ] This change requires a documentation update 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added a test case that will be used to verify my changes 
- [ ] Verified independently on a cluster by reviewer

### CODE REVIEW

File: pkg/products/threescale/reconciler.go [hunk 1]
```diff
@@ -705,10 +705,15 @@ func (r *Reconciler) reconcileExternalDatasources(ctx context.Context, serverCli
 	if err != nil {
 		return integreatlyv1alpha1.PhaseFailed, fmt.Errorf("failed to create redis deletion failure alert for threescale: %w", err)
 	}
-	// create Redis CPU Usage High alert
+	// create Redis Memory Usage High alert
 	if err = resources.CreateRedisMemoryUsageAlerts(ctx, serverClient, r.installation, backendRedis); err != nil {
 		return integreatlyv1alpha1.PhaseFailed, fmt.Errorf("failed to create backend redis prometheus memory usage high alerts for threescale: %s", err)
 	}
+
```
Comment (laurafitzgerald) [inline pkg/products/threescale/reconciler.go:712]: @austincunningham we are doing a refactor which will move some of these places where checks are added to a function in metrics.go, shouldn't change what you're doing here too much but just an fyi. Will ping you on that pr when we have it open.

Comment (laurafitzgerald) [reply pkg/products/threescale/reconciler.go:712]: Here's the pr https://github.com/integr8ly/integreatly-operator/pull/1108

Comment (austincunningham) [reply pkg/products/threescale/reconciler.go:712]: will hold off on this till #1108 is merged, not a blocker for us testing cpu utilization

File: pkg/resources/metrics.go [hunk 2]
```diff
@@ -463,6 +463,70 @@ func CreateRedisMemoryUsageAlerts(ctx context.Context, client k8sclient.Client,
...
+	alertDescription = "Redis Memory Usage is predicted to max with in four hours for instance {{ $labels.instanceID }}. Redis Custom Resource: {{ $labels.resourceID }} in namespace {{ $labels.namespace }} for the product: {{ $labels.productName }}"
+	labels = map[string]string{
+		"severity": "critical",
+	}
+	// building a predict_linear query using 1 hour of data points to predict a 4 hour projection, and checking if it is less than or equal 0
+	//    * [1h] - one hour data points
+	//    * , 4 * 3600 - multiplying data points by 4 hours
+	// and matching by label `job` if the current time is greater than 1 hour of the process start time for the cloud resource operator metrics.
+	//    * on(job) - matching queries by label job across both metrics
+	alertExp = intstr.FromString(fmt.Sprintf("predict_linear(cro_redis_engine_cpu_utilization_average{job='%s'}[1h], 5 * 3600) >= 100 and on(job) (time() - process_start_time_seconds{job='%s'}) / 3600 > 1", job, job))
```
Comment (ciaranRoche) [inline pkg/resources/metrics.go:502]: Just thinking about these prediction alerts on CPU utilization. 

`predict_linear` seems to be used mostly for resource exhaustion (will run out of storage/memory). I am thinking would this lead to noisy alerts when trying to predict based on a counter for CPU utilization :thinking: 

Looking at [AWS docs](https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/CacheMetrics.WhichShouldIMonitor.html) should we be alerting based on a threshold (`RedisCpuUsageHigh`) only :thinking: 

What do you think?

Comment (austincunningham) [reply pkg/resources/metrics.go:502]: I'm not invested in predict_linear this pr is block till https://github.com/integr8ly/integreatly-operator/pull/1108 is merged then will refactor, redis is memory and network bound likely the memory will trigger well before we ever see issues with cpu.

Comment (laurafitzgerald) [reply pkg/resources/metrics.go:502]: For the cpu usage for postgres we just added one alert for over 90% warning for 5 minutes. Seeing as it dropped immediately after any loading I figured we'd just need that one and no predict_linear ones.

Comment (austincunningham) [reply pkg/resources/metrics.go:502]: @laurafitzgerald :thinking: , 5 min might be a bit low, I went with 1 hour which might be too long , think think the sweet spot is some where in between. might go with 15min

**Comment by darahayes** [comment]
We verified the changes together but I cannot give approval on this repo.

**Comment by david-martin** [comment]
@austincunningham If you haven't already, i'd recommend getting an lgtm from cssre.
I'm happy to approve if they are happy

**Comment by austincunningham** [comment]
@david-martin CS-SRE have already approved the SOP for the alert. So I guess they are happy

**Comment by austincunningham** [comment]
/assign @Nanyte25 @brizrobbo @g0ldrush @pat-cremin @pepedocs 
Any of you have time for a review of this

File: pkg/resources/metrics.go [hunk 3]
```diff
@@ -433,10 +440,11 @@ func createRedisMemoryUsageAlerts(ctx context.Context, client k8sclient.Client,
 	job := "cloud-resource-operator-metrics"
 
 	alertName = "RedisMemoryUsageMaxIn4Hours"
-	ruleName = "redis-memory-usage-will-max-in-4-hours"
+	ruleName = fmt.Sprintf("redis-memory-usage-will-max-in-4-hours")
```
Comment (laurafitzgerald) [inline pkg/resources/metrics.go:443]: Do we need the `Sprintf` here?

**Comment by laurafitzgerald** [comment]
@david-martin we have the lgtm from sre and a comment above it was verified by @darahayes prow has suggested yourself as an approver. Would you mind adding the label for us. 🙏

**Pull request merged**